### DISCUSSION TITLE
Unable to Compile my uWebsockets code

### DISCUSSION THREAD

**Discussion by rangpec** [uNetworking/uWebSockets#1814] [Q&A] [answered]
Im trying to learn uWebSockets. Please forgive me for simple questions.
Im using Ubuntu. I installed uWebSockets from 'git clone https://github.com/uNetworking/uWebSockets.git'

after cloning, I did below commands.

1. cd uWebSockets
2. make
3. sudo make install

I tried to compile my code using below compilation command. I am getting errors. Errors shown below.

g++ -o server.o Server.C -I/usr/local/include/uWebSockets -I./uWebSockets/uSockets/src  -L./uWebSockets/uSockets/  -lssl -lcrypto -lz

**Here is my simple server code(Server,C):**

_#include "uWebSockets/src/App.h"
#include <iostream>

int main()
{
        /* Overly simple hello world app */
        int port = std::getenv("SERVER_PORT") ? std::stoi(std::getenv("SERVER_PORT")) : 3000;
        std::cout << "Starting server on port " << port << std::endl;
        uWS::App()
                .get("/*", [port](auto *res, auto * /*req*/) // Capture 'port' in the lambda capture list
                         { res->end("Hello world!"); })
                .listen(port, [port](auto *listen_socket) // Capture 'port' in the lambda capture list
                                {
            if (listen_socket) {
                        std::cout << "Listening on port " << port << std::endl;
            } })
                .run();
        std::cout << "Failed to listen on port 3000" << std::endl;
}_

----------------------
**Im getting below errors. Please help me how to resolve these errors.** 

**$g++ -o server.o Server.C -I/usr/local/include/uWebSockets -I./uWebSockets/uSockets/src -L./uWebSockets/uSockets/ -lssl -lcrypto -lz**
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::wakeupCb(us_loop_t*)':
Server.C:(.text._ZN3uWS4Loop8wakeupCbEP9us_loop_t[_ZN3uWS4Loop8wakeupCbEP9us_loop_t]+0x27): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::preCb(us_loop_t*)':
Server.C:(.text._ZN3uWS4Loop5preCbEP9us_loop_t[_ZN3uWS4Loop5preCbEP9us_loop_t]+0x27): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::postCb(us_loop_t*)':
Server.C:(.text._ZN3uWS4Loop6postCbEP9us_loop_t[_ZN3uWS4Loop6postCbEP9us_loop_t]+0x27): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::init()':
Server.C:(.text._ZN3uWS4Loop4initEv[_ZN3uWS4Loop4initEv]+0x1f): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::create(void*)::{lambda(us_timer_t*)#1}::operator()(us_timer_t*) const':
Server.C:(.text._ZZN3uWS4Loop6createEPvENKUlP10us_timer_tE_clES3_[_ZZN3uWS4Loop6createEPvENKUlP10us_timer_tE_clES3_]+0x2b): undefined reference to `us_timer_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::create(void*)':
Server.C:(.text._ZN3uWS4Loop6createEPv[_ZN3uWS4Loop6createEPv]+0x46): undefined reference to `us_create_loop'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop6createEPv[_ZN3uWS4Loop6createEPv]+0x5e): undefined reference to `us_loop_ext'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop6createEPv[_ZN3uWS4Loop6createEPv]+0x7c): undefined reference to `us_create_timer'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop6createEPv[_ZN3uWS4Loop6createEPv]+0x96): undefined reference to `us_timer_ext'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop6createEPv[_ZN3uWS4Loop6createEPv]+0xc9): undefined reference to `us_timer_set'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::free()':
Server.C:(.text._ZN3uWS4Loop4freeEv[_ZN3uWS4Loop4freeEv]+0x18): undefined reference to `us_loop_ext'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop4freeEv[_ZN3uWS4Loop4freeEv]+0x2f): undefined reference to `us_timer_close'
/usr/bin/ld: Server.C:(.text._ZN3uWS4Loop4freeEv[_ZN3uWS4Loop4freeEv]+0x47): undefined reference to `us_loop_free'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::removePostHandler(void*)':
Server.C:(.text._ZN3uWS4Loop17removePostHandlerEPv[_ZN3uWS4Loop17removePostHandlerEPv]+0x1c): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::removePreHandler(void*)':
Server.C:(.text._ZN3uWS4Loop16removePreHandlerEPv[_ZN3uWS4Loop16removePreHandlerEPv]+0x1c): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::Loop::run()':
Server.C:(.text._ZN3uWS4Loop3runEv[_ZN3uWS4Loop3runEv]+0x18): undefined reference to `us_loop_run'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::create(uWS::Loop*, us_socket_context_options_t)':
Server.C:(.text._ZN3uWS11HttpContextILb0EE6createEPNS_4LoopE27us_socket_context_options_t[_ZN3uWS11HttpContextILb0EE6createEPNS_4LoopE27us_socket_context_options_t]+0x68): undefined reference to `us_create_socket_context'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE6createEPNS_4LoopE27us_socket_context_options_t[_ZN3uWS11HttpContextILb0EE6createEPNS_4LoopE27us_socket_context_options_t]+0x8f): undefined reference to `us_socket_context_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::free()':
Server.C:(.text._ZN3uWS11HttpContextILb0EE4freeEv[_ZN3uWS11HttpContextILb0EE4freeEv]+0x41): undefined reference to `us_socket_context_free'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::listen(char const*, int, int)':
Server.C:(.text._ZN3uWS11HttpContextILb0EE6listenEPKcii[_ZN3uWS11HttpContextILb0EE6listenEPKcii]+0x47): undefined reference to `us_socket_context_listen'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*, int, char*, int)#1}::operator()(us_socket_t*, int, char*, int) const':
Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPciE_clES3_iS4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPciE_clES3_iS4_i]+0x40): undefined reference to `us_socket_timeout'
/usr/bin/ld: Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPciE_clES3_iS4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPciE_clES3_iS4_i]+0x51): undefined reference to `us_socket_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*, int, void*)#1}::operator()(us_socket_t*, int, void*) const':
Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPvE_clES3_iS4_[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tiPvE_clES3_iS4_]+0x37): undefined reference to `us_socket_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*, char*, int)#1}::operator()(us_socket_t*, char*, int) const':
Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i]+0x60): undefined reference to `us_socket_is_shut_down'
/usr/bin/ld: Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i]+0x89): undefined reference to `us_socket_ext'
/usr/bin/ld: Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i]+0x21d): undefined reference to `us_socket_write'
/usr/bin/ld: Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i]+0x231): undefined reference to `us_socket_shutdown'
/usr/bin/ld: Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_i]+0x24f): undefined reference to `us_socket_close'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*, char*, int)#1}::operator()(us_socket_t*, char*, int) const::{lambda(void*, uWS::HttpRequest*)#1}::operator()(void*, uWS::HttpRequest*) const':
Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x3e): undefined reference to `us_socket_timeout'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x4f): undefined reference to `us_socket_ext'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x8e): undefined reference to `us_socket_close'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x1cc): undefined reference to `us_socket_close'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x204): undefined reference to `us_socket_is_closed'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x228): undefined reference to `us_socket_is_shut_down'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvPNS_11HttpRequestEE_clES6_S8_]+0x2f4): undefined reference to `us_socket_timeout'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*, char*, int)#1}::operator()(us_socket_t*, char*, int) const::{lambda(void*, std::basic_string_view<char, std::char_traits<char> >, bool)#1}::operator()(void*, std::basic_string_view<char, std::char_traits<char> >, bool) const':
Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b]+0x66): undefined reference to `us_socket_timeout'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b]+0xbc): undefined reference to `us_socket_timeout'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b]+0x103): undefined reference to `us_socket_is_closed'
/usr/bin/ld: Server.C:(.text._ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b[_ZZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tPciE_clES3_S4_iENKUlPvSt17basic_string_viewIcSt11char_traitsIcEEbE_clES6_SA_b]+0x124): undefined reference to `us_socket_is_shut_down'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()::{lambda(us_socket_t*)#1}::operator()(us_socket_t*) const':
Server.C:(.text._ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tE_clES3_[_ZZN3uWS11HttpContextILb0EE4initEvENKUlP11us_socket_tE_clES3_]+0x52): undefined reference to `us_socket_timeout'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::init()':
Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0x47): undefined reference to `us_socket_context_on_open'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0x72): undefined reference to `us_socket_context_on_close'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0x9d): undefined reference to `us_socket_context_on_data'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0xc8): undefined reference to `us_socket_context_on_writable'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0xf3): undefined reference to `us_socket_context_on_end'
/usr/bin/ld: Server.C:(.text._ZN3uWS11HttpContextILb0EE4initEv[_ZN3uWS11HttpContextILb0EE4initEv]+0x11e): undefined reference to `us_socket_context_on_timeout'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::write(char const*, int, bool, int)':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE5writeEPKcibi[_ZN3uWS11AsyncSocketILb0EE5writeEPKcibi]+0x3f): undefined reference to `us_socket_is_closed'
/usr/bin/ld: Server.C:(.text._ZN3uWS11AsyncSocketILb0EE5writeEPKcibi[_ZN3uWS11AsyncSocketILb0EE5writeEPKcibi]+0xde): undefined reference to `us_socket_write'
/usr/bin/ld: Server.C:(.text._ZN3uWS11AsyncSocketILb0EE5writeEPKcibi[_ZN3uWS11AsyncSocketILb0EE5writeEPKcibi]+0x24f): undefined reference to `us_socket_write'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::getSocketContextData()':
Server.C:(.text._ZN3uWS11HttpContextILb0EE20getSocketContextDataEv[_ZN3uWS11HttpContextILb0EE20getSocketContextDataEv]+0x25): undefined reference to `us_socket_context_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::shutdown()':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE8shutdownEv[_ZN3uWS11AsyncSocketILb0EE8shutdownEv]+0x1d): undefined reference to `us_socket_shutdown'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::close()':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE5closeEv[_ZN3uWS11AsyncSocketILb0EE5closeEv]+0x27): undefined reference to `us_socket_close'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::timeout(unsigned int)':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE7timeoutEj[_ZN3uWS11AsyncSocketILb0EE7timeoutEj]+0x23): undefined reference to `us_socket_timeout'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpResponse<false>::writeMark()':
Server.C:(.text._ZN3uWS12HttpResponseILb0EE9writeMarkEv[_ZN3uWS12HttpResponseILb0EE9writeMarkEv]+0x2c): undefined reference to `us_socket_context'
/usr/bin/ld: Server.C:(.text._ZN3uWS12HttpResponseILb0EE9writeMarkEv[_ZN3uWS12HttpResponseILb0EE9writeMarkEv]+0x39): undefined reference to `us_socket_context_loop'
/usr/bin/ld: Server.C:(.text._ZN3uWS12HttpResponseILb0EE9writeMarkEv[_ZN3uWS12HttpResponseILb0EE9writeMarkEv]+0x41): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::getSocketContextDataS(us_socket_t*)':
Server.C:(.text._ZN3uWS11HttpContextILb0EE21getSocketContextDataSEP11us_socket_t[_ZN3uWS11HttpContextILb0EE21getSocketContextDataSEP11us_socket_t]+0x25): undefined reference to `us_socket_context_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::getAsyncSocketData()':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE18getAsyncSocketDataEv[_ZN3uWS11AsyncSocketILb0EE18getAsyncSocketDataEv]+0x1d): undefined reference to `us_socket_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::AsyncSocket<false>::getLoopData()':
Server.C:(.text._ZN3uWS11AsyncSocketILb0EE11getLoopDataEv[_ZN3uWS11AsyncSocketILb0EE11getLoopDataEv]+0x1d): undefined reference to `us_socket_context'
/usr/bin/ld: Server.C:(.text._ZN3uWS11AsyncSocketILb0EE11getLoopDataEv[_ZN3uWS11AsyncSocketILb0EE11getLoopDataEv]+0x2a): undefined reference to `us_socket_context_loop'
/usr/bin/ld: Server.C:(.text._ZN3uWS11AsyncSocketILb0EE11getLoopDataEv[_ZN3uWS11AsyncSocketILb0EE11getLoopDataEv]+0x32): undefined reference to `us_loop_ext'
/usr/bin/ld: /tmp/ccapRFH7.o: in function `uWS::HttpContext<false>::getSocketContext(us_socket_t*)':
Server.C:(.text._ZN3uWS11HttpContextILb0EE16getSocketContextEP11us_socket_t[_ZN3uWS11HttpContextILb0EE16getSocketContextEP11us_socket_t]+0x1d): undefined reference to `us_socket_context'
collect2: error: ld returned 1 exit status
$

**If Im missing any steps or need to include any directory or library while compiling, please help me with exact compilation command.**

**Comment by slavslavov** [comment]
uSockets is a git submodule and you need to pull it before you can compule

**Reply by rangpec** [reply]
Using below command I cloned. I got above errors after doing this way of clone.

_git clone --recurse-submodules https://github.com/uNetworking/uWebSockets.git_

I could see the uSockets directory.

$ ls -lrt
total 1780
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 misc
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 libEpollBenchmarker
drwxr-xr-x 4 rp rp   4096 Nov 25 16:55 fuzzing
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 cluster
-rw-r--r-- 1 rp rp    552 Nov 25 16:55 build.h
-rw-r--r-- 1 rp rp   3678 Nov 25 16:55 build.c
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 benchmarks
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 autobahn
-rw-r--r-- 1 rp rp   6101 Nov 25 16:55 README.md
-rw-r--r-- 1 rp rp    640 Nov 25 16:55 Makefile
-rw-r--r-- 1 rp rp  11357 Nov 25 16:55 LICENSE
-rw-r--r-- 1 rp rp    377 Nov 25 16:55 GNUmakefile
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 tests
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 src
drwxr-xr-x 2 rp rp   4096 Nov 25 16:55 h1spec
-rwxr-xr-x 1 rp rp  16448 Nov 25 16:58 build
-rwxr-xr-x 1 rp rp 111120 Nov 25 16:58 EchoBody
-rwxr-xr-x 1 rp rp 128424 Nov 25 16:58 HelloWorldThreaded
-rwxr-xr-x 1 rp rp  16032 Nov 25 16:58 Http3Server
-rwxr-xr-x 1 rp rp 192512 Nov 25 16:58 Broadcast
-rwxr-xr-x 1 rp rp 121440 Nov 25 16:58 HelloWorld
-rwxr-xr-x 1 rp rp 134208 Nov 25 16:58 Crc32
-rwxr-xr-x 1 rp rp 127624 Nov 25 16:59 ServerName
-rwxr-xr-x 1 rp rp 167624 Nov 25 16:59 EchoServer
-rwxr-xr-x 1 rp rp 196976 Nov 25 16:59 BroadcastingEchoServer
-rwxr-xr-x 1 rp rp 187744 Nov 25 16:59 UpgradeSync
-rwxr-xr-x 1 rp rp 189192 Nov 25 16:59 UpgradeAsync
-rwxr-xr-x 1 rp rp 125784 Nov 25 16:59 ParameterRoutes
drwxr-xr-x 3 rp rp   4096 Nov 25 17:08 examples
drwxr-xr-x 9 rp rp   4096 Nov 25 17:24 uSockets
$
$
$ cd uSockets/
$ ls -lrt
total 576
drwxr-xr-x  2 rp rp   4096 Nov 25 16:55 tests
drwxr-xr-x  6 rp rp   4096 Nov 25 16:55 src
-rw-r--r--  1 rp rp     58 Nov 25 16:55 module.modulemap
drwxr-xr-x  2 rp rp   4096 Nov 25 16:55 misc
drwxr-xr-x  3 rp rp   4096 Nov 25 16:55 examples
-rw-r--r--  1 rp rp   1509 Nov 25 16:55 README.md
-rw-r--r--  1 rp rp   3454 Nov 25 16:55 Makefile
-rw-r--r--  1 rp rp  11357 Nov 25 16:55 LICENSE
drwxr-xr-x 12 rp rp   4096 Nov 25 16:56 boringssl
drwxr-xr-x  9 rp rp   4096 Nov 25 16:56 lsquic
-rw-r--r--  1 rp rp  51392 Nov 25 17:24 bsd.o
-rw-r--r--  1 rp rp  37504 Nov 25 17:24 context.o
-rw-r--r--  1 rp rp  30328 Nov 25 17:24 loop.o
-rw-r--r--  1 rp rp   2712 Nov 25 17:24 quic.o
-rw-r--r--  1 rp rp  28664 Nov 25 17:24 socket.o
-rw-r--r--  1 rp rp  17896 Nov 25 17:24 udp.o
-rw-r--r--  1 rp rp  31824 Nov 25 17:24 epoll_kqueue.o
-rw-r--r--  1 rp rp   2712 Nov 25 17:24 gcd.o
-rw-r--r--  1 rp rp   2712 Nov 25 17:24 libuv.o
-rw-r--r--  1 rp rp   2712 Nov 25 17:24 openssl.o
-rw-r--r--  1 rp rp   2720 Nov 25 17:24 io_context.o
-rw-r--r--  1 rp rp   2720 Nov 25 17:24 io_socket.o
-rw-r--r--  1 rp rp   2712 Nov 25 17:24 io_loop.o
-rw-r--r--  1 rp rp 302144 Nov 25 17:24 uSockets.a
$

**Comment by slavslavov** [comment]
Sorry, you already said you have installed it, so it is not the submodule
causing the problem

**Comment by uNetworkingAB** [comment, marked as answer]
If you simply hit "make", you will see exactly how the examples are compiled:

`g++  -march=native -O3 -Wpedantic -Wall -Wextra -Wsign-conversion -Wconversion -std=c++20 -Isrc -IuSockets/src -flto examples/EchoBody.cpp  uSockets/*.o -lz -o EchoBody`

**Reply by uNetworkingAB** [reply]
You are missing `uSockets/*.o` or any such alternative

**Reply by rangpec** [reply]
Hi,

Your answer helped me. 

**Below is my modified compilation command which compiles successfully:**

_g++  -march=native -O3 -Wpedantic -Wall -Wextra -Wsign-conversion -Wconversion -std=c++20 -IuWebSockets/src -IuWebSockets/uSockets/src -flto Server.C  uWebSockets/uSockets/*.o -lz -o server.o_

Thanks for helping me. 

Also please do another help. I need a very document to read and understand uWebSockets. Could you please share me?
If any weblink also fine.

Advance Thanks!!

**Comment by rangpec** [comment]
@uNetworkingAB could you please give me very good documentation for learning uWebsockets which Systematically explains all concepts and sample C++ program to execute it.
Im new to this websocket and uWebsocket. Sorry for disturbing you. Please point me to right weblink or document.

Thanks,
Ranganadhan

**Reply by MacroMelon** [reply]
Hello! You might have already seen these, but I found them extremely useful:

(From the main readme) 

"Start building your Http & WebSocket apps in no time; <a href="https://github.com/uNetworking/uWebSockets/blob/master/misc/READMORE.md">read the user manual</a> and <a href="https://github.com/uNetworking/uWebSockets/tree/master/examples">see examples</a>. You can browse our <a href="https://unetworking.github.io/uWebSockets.js/generated/">TypeDoc</a> for a quick overview."

Good luck and happy programming :)

**Comment by rangpec** [comment]
ok Thank you very much.

**Comment by rhildred** [comment]
In 2025 it is also necessary, in my experience to add `-lssl -lcrypto`

For instance:

```
g++ -march=native -O3 -Wpedantic -Wall -Wextra -Wsign-conversion -Wconversion -std=c++20 -Isrc -IuSockets/src -flto examples/HttpServer.cpp uSockets/*.o -lz -lssl -lcrypto -o HttpServer
```

Hope that this helps.

**Reply by uNetworkingAB** [reply]
If you read the readme, the front page, there is an example how to build with SSL.
